### ISSUE TITLE
"Unknown error inside the compiler" and OutOfMemoryError

### ISSUE THREAD

**Issue by mark-fathom** [processing/processing#3929]
Seen on Mac OS X:
```
Exception in thread "TimerQueue" 2015-09-30 14:23:06.138 Processing[1729:508902] java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.addWaiter(AbstractQueuedSynchronizer.java:606)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
	at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
	at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
	at java.awt.EventQueue.wakeup(EventQueue.java:1330)
	at java.awt.EventQueue.access$100(EventQueue.java:97)
	at java.awt.EventQueue$2.wakeup(EventQueue.java:213)
	at sun.awt.SunToolkit.wakeupEventQueue(SunToolkit.java:309)
	at sun.awt.PostEventQueue.postEvent(SunToolkit.java:2177)
	at sun.awt.SunToolkit.postEvent(SunToolkit.java:507)
	at sun.lwawt.LWToolkit.postEvent(LWToolkit.java:466)
	at sun.lwawt.LWComponentPeer.postEvent(LWComponentPeer.java:1164)
	at sun.lwawt.LWWindowPeer.postMouseExitedEvent(LWWindowPeer.java:907)
	at sun.lwawt.LWWindowPeer.notifyMouseEvent(LWWindowPeer.java:727)
	at sun.lwawt.macosx.CPlatformResponder.handleMouseEvent(CPlatformResponder.java:83)
	at sun.lwawt.macosx.CPlatformView.deliverMouseEvent(CPlatformView.java:197)
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
2015-09-30 14:23:06.145 Processing[1729:508902] (
	0   CoreFoundation                      0x00007fff925a503c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff8cb4476e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff925a4bd9 -[NSException raise] + 9
	3   JavaNativeFoundation                0x000000011d0d5539 JNFCallVoidMethod + 209
	4   libawt_lwawt.dylib                  0x000000011f2e8041 -[AWTView deliverJavaMouseEvent:] + 755
	5   AppKit                              0x00007fff95b9019a -[NSTrackingArea _dispatchMouseExited:] + 75
	6   AppKit                              0x00007fff95b900b3 -[NSTrackingArea _mouseExited:] + 95
	7   AppKit                              0x00007fff95ab2797 -[NSApplication sendEvent:] + 3917
	8   libosxapp.dylib                     0x000000011f3973ac -[NSApplicationAWT sendEvent:] + 265
	9   AppKit                              0x00007fff959dbb68 -[NSApplication run] + 711
	10  libosxapp.dylib                     0x000000011f39723e +[NSApplicationAWT runAWTLoopWithApp:] + 156
	11  libawt_lwawt.dylib                  0x000000011f32554b -[AWTStarter starter:] + 905
	12  Foundation                          0x00007fff8be2bdd0 __NSThreadPerformPerform + 293
	13  CoreFoundation                      0x00007fff924c1a01 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	14  CoreFoundation                      0x00007fff924b3b8d __CFRunLoopDoSources0 + 269
	15  CoreFoundation                      0x00007fff924b31bf __CFRunLoopRun + 927
	16  CoreFoundation                      0x00007fff924b2bd8 CFRunLoopRunSpecific + 296
	17  libjli.dylib                        0x0000000107f798fb CreateExecutionEnvironment + 871
	18  libjli.dylib                        0x0000000107f75644 JLI_Launch + 1952
	19  Processing                          0x0000000107f0e8ab launch + 5243
	20  Processing                          0x0000000107f0d2e6 main + 102
	21  Processing                          0x0000000107f0d274 start + 52
)
	at processing.mode.java.Compiler.compile(Compiler.java:118)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:159)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:131)
	at processing.mode.java.JavaMode.handleLaunch(JavaMode.java:151)
	at processing.mode.java.JavaEditor$35.run(JavaEditor.java:1087)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at org.eclipse.jdt.internal.compiler.classfmt.ClassFileStruct.utf8At(ClassFileStruct.java:60)
	at org.eclipse.jdt.internal.compiler.classfmt.MethodInfo.createMethod(MethodInfo.java:50)
	at org.eclipse.jdt.internal.compiler.classfmt.ClassFileReader.<init>(ClassFileReader.java:278)
	at org.eclipse.jdt.internal.compiler.classfmt.ClassFileReader.<init>(ClassFileReader.java:138)
	at org.eclipse.jdt.internal.compiler.classfmt.ClassFileReader.read(ClassFileReader.java:116)
	at org.eclipse.jdt.internal.compiler.classfmt.ClassFileReader.read(ClassFileReader.java:104)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathJar.findClass(ClasspathJar.java:98)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.findClass(FileSystem.java:262)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.findType(FileSystem.java:354)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.askForType(LookupEnvironment.java:158)
	at org.eclipse.jdt.internal.compiler.lookup.PackageBinding.getType(PackageBinding.java:145)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.findType(Scope.java:2038)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.getTypeOrPackage(Scope.java:3350)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.getType(Scope.java:3042)
	at org.eclipse.jdt.internal.compiler.ast.SingleTypeReference.getTypeBinding(SingleTypeReference.java:51)
	at org.eclipse.jdt.internal.compiler.ast.TypeReference.internalResolveType(TypeReference.java:472)
	at org.eclipse.jdt.internal.compiler.ast.TypeReference.resolveType(TypeReference.java:571)
	at org.eclipse.jdt.internal.compiler.ast.TypeReference.resolveType(TypeReference.java:567)
	at org.eclipse.jdt.internal.compiler.ast.TypeReference.resolveSuperType(TypeReference.java:541)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.findSupertype(ClassScope.java:1287)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.connectSuperclass(ClassScope.java:938)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.connectTypeHierarchy(ClassScope.java:1095)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.connectTypeHierarchy(CompilationUnitScope.java:323)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.completeTypeBindings(LookupEnvironment.java:227)
	at org.eclipse.jdt.internal.compiler.Compiler.internalBeginToCompile(Compiler.java:746)
	at org.eclipse.jdt.internal.compiler.Compiler.beginToCompile(Compiler.java:383)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:428)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4129)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1694)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1385)
	at org.eclipse.jdt.core.compiler.batch.BatchCompiler.compile(BatchCompiler.java:80)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Exception in thread "AWT-EventQueue-0" java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.addConditionWaiter(AbstractQueuedSynchronizer.java:1855)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2068)
	at javax.swing.TimerQueue.run(TimerQueue.java:192)
	at java.lang.Thread.run(Thread.java:745)
processing.app.SketchException: Unknown error inside the compiler.
	at processing.mode.java.Compiler.compile(Compiler.java:121)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:159)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:131)
	at processing.mode.java.JavaMode.handleLaunch(JavaMode.java:151)
	at processing.mode.java.JavaEditor$35.run(JavaEditor.java:1087)
	at java.lang.Thread.run(Thread.java:745)
java.lang.OutOfMemoryError: Java heap space
	at java.awt.EventQueue.setCurrentEventAndMostRecentTimeImpl(EventQueue.java:1214)
	at java.awt.EventQueue.setCurrentEventAndMostRecentTime(EventQueue.java:1205)
	at java.awt.Component.dispatchEventImpl(Component.java:4727)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:995)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:450)
	at java.awt.Component.dispatchEventImpl(Component.java:4752)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

and a second run produced:

```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at processing.mode.java.Compiler.compile(Compiler.java:118)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:159)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:131)
	at processing.mode.java.JavaMode.handleLaunch(JavaMode.java:151)
	at processing.mode.java.JavaEditor$35.run(JavaEditor.java:1087)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
processing.app.SketchException: Unknown error inside the compiler.
	at processing.mode.java.Compiler.compile(Compiler.java:121)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:159)
	at processing.mode.java.JavaBuild.build(JavaBuild.java:131)
	at processing.mode.java.JavaMode.handleLaunch(JavaMode.java:151)
	at processing.mode.java.JavaEditor$35.run(JavaEditor.java:1087)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by benfry** [comment]
Seen in person, but not repeatable. Keeping an eye on it.

**Comment by JakubValtar** [comment]
Fixed in #3934. Closed Editors were not garbage collected.
